eth/signer.go

```diff
@@ -132,24 +139,92 @@ func (s *Signer) CompactSign(buf []byte) ([]byte, error) {
 	return rsv[:64], nil
 }
 
-// SignWithNonce generates a 32-byte nonce with crypto/rand and returns
-// s.CompactSign(append(buf, nonce)). This can be verified in Solidity
-// via ecrecover with a hash value of
-// keccak256(abi.encodePacked(buf,nonce)).
-func (s *Signer) SignWithNonce(buf []byte) ([]byte, [32]byte, error) {
+// AppendRandomNonce appends random 32 bytes to the buffer, commonly used in
+// signature nonces.
+func AppendRandomNonce(buf []byte) ([]byte, [32]byte, error) {
 	var nonce [32]byte
 	if n, err := rand.Read(nonce[:]); n != 32 || err != nil {
 		return nil, nonce, fmt.Errorf("read 32 random bytes: got %d bytes with err %v", n, err)
 	}
 
-	sig, err := s.CompactSign(append(buf, nonce[:]...))
+	return append(buf, nonce[:]...), nonce, nil
+}
+
+// ToEthSignedMessageHash converts a given message to conform to the signed data
+// standard according to EIP-191.
+func ToEthPersonalSignedMessage(message []byte) []byte {
+	prefix := []byte(fmt.Sprintf("\x19Ethereum Signed Message:\n%d", len(message)))
+	return append(prefix, message...)
+}
+
+// PersonalSign returns an EIP-191 conform personal ECDSA signature of buf
+// Convenience wrapper for s.CompactSign(toEthPersonalSignedMessage(buf))
+func (s *Signer) PersonalSign(buf []byte) ([]byte, error) {
+	sig, _, err := s.sign(buf, signOpts{
+		raw:       false,
+		compact:   true,
+		personal:  true,
+		withNonce: false,
+	})
+	return sig, err
+}
+
+// PersonalSignWithNonce generates a 32-byte nonce with crypto/rand and returns
+// s.PersonalSign(append(buf, nonce)).
+func (s *Signer) PersonalSignWithNonce(buf []byte) ([]byte, [32]byte, error) {
+	sig, nonce, err := s.sign(buf, signOpts{
+		raw:       false,
+		compact:   true,
+		personal:  true,
+		withNonce: true,
+	})
+	return sig, *nonce, err
+
+}
+
+type signOpts struct {
+	raw, compact, personal, withNonce bool
+}
+
+// sign signs a given buffer depending on the chosen options:
+// withNonce = true, appends a nonce to the message
+// compact = true, returns a compactified version of the signature according to
+// EIP-2098.
+// personal = true, adds a prefix to the message to conform to the EIP-191
+// personal message standard.
+// raw = false, the message is hashed before signing
+func (s *Signer) sign(buf []byte, opts signOpts) ([]byte, *[32]byte, error) {
+	var nonce [32]byte
+	var err error
+
+	if opts.withNonce {
+		buf, nonce, err = AppendRandomNonce(buf)
+	}
+
+	if opts.personal {
+		buf = ToEthPersonalSignedMessage(buf)
+	}
+
+	if !opts.raw {
+		buf = crypto.Keccak256(buf)
+	}
+
+	sig, err := crypto.Sign(buf, s.key)
 	if err != nil {
-		return nil, nonce, err
+		return nil, nil, err
 	}
-	return sig, nonce, nil
+
+	if opts.compact {
```

Comment:
(nit) If you return early then you can reduce nesting and improve readability: https://testing.googleblog.com/2017/06/code-health-reduce-nesting-reduce.html

```
if !opts.compact {
  return sig, nonce, err
}

// compact
```

Reply:
done